File: JaeHunGang/4주차/BJ_11725_트리의_부모_찾기.js
// https://www.acmicpc.net/problem/11725
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const nodes = parseInt(input[0], 10);
const connectionNodes = input.splice(1).map((line) => {
  const [node1, node2] = line.split(" ").map((v) => parseInt(v, 10));
  return [node1, node2];
});
const parentNode = Array.from({ length: nodes + 1 }, () => []);
const result = Array(nodes + 1).fill(0);

for (const [node1, node2] of connectionNodes) {
  parentNode[node1].push(node2);
  parentNode[node2].push(node1);
}

function DFS(current, parent) {
  result[current] = parent;

  for (const next of parentNode[current]) {
    if (result[next] === 0) {
      DFS(next, current);
    }
  }
}

DFS(1, 0);
console.log(result.splice(2).join("\n"));

File: JeongHunHui/4주차/BJ_11725_트리의_부모_찾기.py
# 16:38
from collections import deque
n = int(input())
nodes = [[] for _ in range(n+1)]
is_visited = [False for _ in range(n+1)]
for _ in range(n-1):
  node1, node2 = map(int, input().split())
  nodes[node1].append(node2)
  nodes[node2].append(node1)

is_visited[1] = True
dq = deque([])
for node in nodes[1]:
  dq.append((node, 1))
  is_visited[node] = True

answers = [None]*(n+1)
while dq:
  node, parent_node = dq.popleft()
  answers[node] = parent_node
  for target_node in nodes[node]:
    if not is_visited[target_node]:
      is_visited[target_node] = True
      dq.append((target_node, node))

for answer in answers[2:]:
  print(answer)